mlflow/store/artifact/mlflow_artifacts_repo.py

```diff
@@ -0,0 +1,81 @@
+from urllib.parse import urlparse
+import posixpath
+from collections import namedtuple
+import requests
+
+from mlflow.store.artifact.http_artifact_repo import HttpArtifactRepository
+from mlflow.tracking._tracking_service.utils import get_tracking_uri
+from mlflow.exceptions import MlflowException
+
+
+def _parse_artifact_uri(artifact_uri):
+    ParsedURI = namedtuple("ParsedURI", "scheme host port path")
+    parsed_uri = urlparse(artifact_uri)
+    return ParsedURI(parsed_uri.scheme, parsed_uri.hostname, parsed_uri.port, parsed_uri.path)
+
+
+def _check_if_host_is_numeric(hostname):
+    if hostname:
+        try:
+            float(hostname)
+            return True
+        except ValueError:
+            return False
+    else:
+        return False
+
+
+def _validate_port_mapped_to_hostname(uri_parse):
+    # This check is to catch an mlflow-artifacts uri that has a port designated but no
+    # hostname specified. `urllib.parse.urlparse` will treat such a uri as a filesystem
+    # definition, mapping the provided port as a hostname value if this condition is not
+    # validated.
+    if uri_parse.host and _check_if_host_is_numeric(uri_parse.host) and not uri_parse.port:
+        raise MlflowException(
+            f"The mlflow-artifacts uri was supplied with a port number: "
+            f"{uri_parse.host}, but no host was defined."
+        )
+
+
+class MlflowArtifactsRepository(HttpArtifactRepository):
+    """Scheme wrapper around HttpArtifactRepository for mlflow-artifacts server functionality"""
+
+    def __init__(self, artifact_uri):
+
+        self._session = requests.Session()
```

Comment:
Do we need this line? `HttpArtifactRepository.__init__` does the same thing, right?

Reply:
I was wondering the same thing. When I don't create a requests.Session() object, I get this Exception (that is ignored in the test suite, but it throws nonetheless):

```Traceback (most recent call last):
  File "/Users/benwilson/Repos/mlflow/mlflow/store/artifact/http_artifact_repo.py", line 18, in __del__
    self._session.close()
AttributeError: 'MlflowArtifactsRepository' object has no attribute '_session'
```

Reply:
Interesting. It looks like `__del__` even when `__init__` fails.

Reply:
`__del__` of `HttpArtifactRepository` should look like:

```
del __del__(self):
    if hasattr(self. "_session"):
        self.close()
```

Reply:
I think it's because requests keeps a session alive by default https://2.python-requests.org/en/master/user/advanced/#keep-alive so that call to __del__ is the only thing saving our connection pool from exhaustion ;) I think the other way to 'auto-handle it' is by wrapping the request call in a with statement, but that really wouldn't work with how HttpArtifactRepository needs to function.
The interesting thing about it is that the subclass of MlflowArtifactsRepository can't inherit the `requests.Session()` instantiation from `super()`. It will throw saying that it's not accessible. Must be some sort of protection in requests that keep it instance-scoped to protect people from leaving sessions active inadvertently?

Reply:
great point! Making that validation check and I'll push in a minute :)